analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs

```diff
@@ -185,12 +196,42 @@ private IEnumerable<Compilation> CompileRazor()
             {
                 Directory.CreateDirectory(tempPath);
 
-                foreach (var file in Directory.GetFiles("TestFramework\\Razor\\EmptyProject").Concat(builder.Paths.Select(TestCasePath)))
+                List<string> languages = new();
+                if (builder.ParseOptions != null && builder.ParseOptions.Any())
+                {
+                    foreach (var parseOption in builder.ParseOptions)
+                    {
+                        if (parseOption is CSharpParseOptions csharpParseOptions)
+                        {
+                            languages.Add(GetLanguageVersionReference(csharpParseOptions));
+                        }
+                    }
+                }
+                else
                 {
-                    File.Copy(file, Path.Combine(tempPath, Path.GetFileName(file)));
+                    languages.Add("latest");
                 }
 
-                return new[] { workspace.OpenProjectAsync(Path.Combine(tempPath, "EmptyProject.csproj")).Result.GetCompilationAsync().Result };
+                foreach (var lang in languages)
+                {
+                    Directory.CreateDirectory(Path.Combine(tempPath, lang));
+                    // Copy all the files
+                    foreach (var file in Directory.GetFiles("TestFramework\\Razor\\EmptyProject").Concat(builder.Paths.Select(TestCasePath)))
+                    {
+                        File.Copy(file, Path.Combine(tempPath, lang, Path.GetFileName(file)));
+                    }
+                    var csprojPath = Path.Combine(tempPath, lang, "EmptyProject.csproj");
+                    var destinationXml = XElement.Load(csprojPath);
+
+                    // Set TargetFramework
+                    destinationXml.Descendants("TargetFramework").Single().Value = builder.RazorFramework;
```

Comment:
Setting the target framework (and later the project references) directly to the copies, instead of doing it in the original `EmptyProject.csproj`. This way we will guarantee that the original keeps the default values without being overridden with a new version targeting the new framework for every test (same for project references). 
Moreover, opening the original csproj, editing its properties, and saving it might be even slower than editing a property of multiple, already loaded .csproj files.